server/service/src/stocktake/update.rs

```diff
@@ -194,252 +217,214 @@ struct StocktakeGenerateJob {
-        )
-    } else {
-        None
-    };
+    let delta = counted_number_of_packs - row.snapshot_number_of_packs;
+
+    let stock_line_row = stock_line.to_owned();
+
+    let pack_size = row.pack_size.unwrap_or(stock_line_row.pack_size);
+    let expiry_date = row.expiry_date.or(stock_line_row.expiry_date);
+    let cost_price_per_pack = row
+        .cost_price_per_pack
+        .unwrap_or(stock_line_row.cost_price_per_pack);
+    let sell_price_per_pack = row
+        .sell_price_per_pack
+        .unwrap_or(stock_line_row.sell_price_per_pack);
+
+    // If no change in stock quantity, we just update the stock line (no inventory adjustment)
+    if delta == 0.0 {
+        let updated_stock_line = StockLineRow {
+            location_id: row.location_id,
+            batch: row.batch,
+            pack_size,
+            cost_price_per_pack,
+            sell_price_per_pack,
+            expiry_date,
+            ..stock_line_row
+        }
+        .to_owned();
 
-    let updated_line = StockLineRow {
-        id: stock_line.id.clone(),
-        item_link_id: stock_line_item_id.clone(),
-        store_id: stock_line.store_id.clone(),
-        location_id: stocktake_line.line.location_id.clone(),
-        batch: stocktake_line.line.batch.clone(),
-        pack_size: stocktake_line
-            .line
-            .pack_size
-            .unwrap_or(stock_line.pack_size),
-        cost_price_per_pack: stocktake_line
-            .line
-            .cost_price_per_pack
-            .unwrap_or(stock_line.cost_price_per_pack),
-        sell_price_per_pack: stocktake_line
-            .line
-            .sell_price_per_pack
-            .unwrap_or(stock_line.sell_price_per_pack),
-        // TODO might get negative!
-        available_number_of_packs: stock_line.available_number_of_packs + delta,
-        total_number_of_packs: stock_line.total_number_of_packs + delta,
-        expiry_date: stocktake_line.line.expiry_date.or(stock_line.expiry_date),
-        on_hold: stock_line.on_hold,
-        note: stock_line.note.clone(),
-        supplier_link_id: stock_line_supplier_id,
-        barcode_id: stock_line.barcode_id.clone(),
+        return Ok(StockLineJob {
+            stock_in_out_or_update: Some(StockChange::StockUpdate(updated_stock_line)),
+            stocktake_line: None,
+            location_movement: None,
+            update_inventory_adjustment_reason: None,
+        });
     };
 
-    let stock_line_item =
-        match ItemRowRepository::new(connection).find_active_by_id(&stock_line_item_id)? {
-            Some(item) => item,
-            None => {
-                return Err(UpdateStocktakeError::InternalError(format!(
-                    "Can't find item {} for existing stocktake line {}!",
-                    &stock_line_item_id, stocktake_line.line.id
-                )))
+    let quantity_change = f64::abs(delta);
+    let invoice_line_id = uuid();
+
+    let update_inventory_adjustment_reason =
```

Comment:
ah yeah nice one - [3cf4418](https://github.com/msupply-foundation/open-msupply/pull/3933/commits/3cf4418b4ffbf84ea49c2122726ee92205b67812)

Reply:
thanks, but I thought it would be nice to separate it in `generate` method, upon trying, it didn't quite turn out that nice. Thanks for the change though